src/Donors/Controllers/DonorRequestController.php

```diff
@@ -0,0 +1,173 @@
+<?php
+
+namespace Give\Donors\Controllers;
+
+use Give\Donations\ValueObjects\DonationMetaKeys;
+use Give\Donors\Models\Donor;
+use Give\Donors\ValueObjects\DonorRoute;
+use Give\Framework\QueryBuilder\JoinQueryBuilder;
+use Give\Framework\QueryBuilder\QueryBuilder;
+use WP_Error;
+use WP_REST_Request;
+use WP_REST_Response;
+
+/**
+ * @unreleased
+ */
+class DonorRequestController
+{
+    /**
+     * @unreleased
+     *
+     * @return WP_Error | WP_REST_Response
+     */
+    public function getDonor(WP_REST_Request $request)
+    {
+        $donor = Donor::find($request->get_param('id'));
+
+        if ( ! $donor) {
+            return new WP_Error('donor_not_found', __('Donor not found', 'give'), ['status' => 404]);
+        }
+
+        return new WP_REST_Response($this->escDonor($donor));
+    }
+
+    /**
+     * @unreleased
+     */
+    public function getDonors(WP_REST_Request $request): WP_REST_Response
+    {
+        $page = $request->get_param('page');
+        $perPage = $request->get_param('per_page');
+        $sortColumn = $this->getSortColumn($request->get_param('sort'));
+        $sortDirection = $request->get_param('direction');
+
+        $query = Donor::query();
+
+        // Donors only can be donors if they have donations associated with them
+        if ($request->get_param('onlyWithDonations')) {
+            $query->join(function (JoinQueryBuilder $builder) {
+                // The donationmeta1.donation_id should be used in other "donationmeta" joins to make sure we are retrieving data from the proper donation
+                $builder->innerJoin('give_donationmeta', 'donationmeta1')
+                    ->joinRaw("ON donationmeta1.meta_key = '" . DonationMetaKeys::DONOR_ID . "' AND donationmeta1.meta_value = ID");
+
+                // Include only current payment "mode"
+                $mode = give_is_test_mode() ? 'test' : 'live';
+                $builder->innerJoin('give_donationmeta', 'donationmeta2')
+                    ->joinRaw("ON donationmeta2.meta_key = '" . DonationMetaKeys::MODE . "' AND donationmeta2.meta_value = '{$mode}' AND donationmeta2.donation_id = donationmeta1.donation_id");
+            });
+
+
+            if ($campaignId = $request->get_param('campaignId')) {
+                // Filter by CampaignId - Donors only can be filtered by campaignId if they donated to a campaign
+                $query->join(function (JoinQueryBuilder $builder) use ($campaignId) {
+                    $builder->innerJoin('give_donationmeta', 'donationmeta3')
+                        ->joinRaw("ON donationmeta3.meta_key = '" . DonationMetaKeys::CAMPAIGN_ID . "' AND donationmeta3.meta_value = {$campaignId} AND donationmeta3.donation_id = donationmeta1.donation_id");
+                });
+            }
+
+            if ($request->get_param('hideAnonymousDonors')) {
```

Comment:
See other comment on keeping parameters positive.

Reply:
Resolved in ded974c38462d50bb6a55d9a2d2e9aa5a7a6aa8e